src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -4200,6 +4203,38 @@ private IOperation ConvertToIDisposable(IOperation operand, ITypeSymbol iDisposa
         {
             StartVisitingStatement(operation);
 
+            // `lock (l) { }` on value of type `System.Threading.Lock` is lowered to `using (l.EnterLockScope()) { }`.
+            if (operation.LockedValue.Type?.IsLockType() == true)
+            {
+                var lockTypeInfo = operation.LockedValue.Type.TryFindLockTypeInfo();
+
+                var lockObject = VisitRequired(operation.LockedValue);
+
```

Comment:
This looks suspicious. I think this value should be visited in the environment (the stack frame, etc) that ```HandleUsingOperationParts``` sets up for visiting resources. Perhaps instead of passing a boollean flag, we could pass a delegate that wraps the resource into the enter lock scope call after the resource is visited by ```HandleUsingOperationParts```?